File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/request/RequestCtxData.java
package ru.citeck.ecos.records3.record.request;

import lombok.AllArgsConstructor;
import lombok.Getter;
import ru.citeck.ecos.commons.utils.MandatoryParam;
import ru.citeck.ecos.commons.utils.StringUtils;
import ru.citeck.ecos.records3.record.request.msg.MsgLevel;

import java.util.*;

@Getter
@AllArgsConstructor
public class RequestCtxData<T> {

    private final String requestId;
    private final List<String> requestTrace;
    private final Locale locale;
    private final Map<String, Object> ctxAtts;
    private final boolean computedAttsDisabled;
    private final MsgLevel msgLevel;

    public static <T> Builder<T> create() {
        return new Builder<>();
    }

    public Builder<T> modify() {
        return new Builder<>(this);
    }

    public static class Builder<T> {

        private String requestId;
        private List<String> requestTrace;
        private Locale locale = Locale.ENGLISH;

        private Map<String, Object> contextAttributes = Collections.emptyMap();

        private boolean compAttsDisabled;
        private MsgLevel msgLevel = MsgLevel.WARN;

        public Builder() {
        }

        public Builder(RequestCtxData<T> base) {
            this.requestId = base.getRequestId();
            this.requestTrace = base.getRequestTrace();
            this.locale = base.getLocale();
            this.contextAttributes = base.ctxAtts;
            this.compAttsDisabled = base.computedAttsDisabled;
            this.msgLevel = base.msgLevel;
        }

        public Builder<T> merge(RequestCtxData<T> other) {
            this.contextAttributes = new HashMap<>(this.contextAttributes);
            other.getCtxAtts().forEach((k, v) -> this.contextAttributes.put(k, v));
            this.compAttsDisabled = other.computedAttsDisabled;
            return this;
        }

        public Builder<T> setMsgLevel(MsgLevel msgLevel) {
            this.msgLevel = msgLevel;
            return this;
        }

        public Builder<T> setCompAttsDisabled(boolean computedAttsDisabled) {
            this.compAttsDisabled = computedAttsDisabled;
            return this;
        }

        public Builder<T> setCtxAtts(Map<String, Object> contextAttributes) {
            this.contextAttributes = contextAttributes;
            return this;
        }

        public Builder<T> setRequestId(String requestId) {
            this.requestId = requestId;
            return this;
        }

        public Builder<T> setRequestTrace(List<String> requestTrace) {
            this.requestTrace = requestTrace;
            return this;
        }

        public Builder<T> setLocale(Locale locale) {
            this.locale = locale;
            return this;
        }

        public RequestCtxData<T> build() {

            MandatoryParam.check("locale", locale);

            if (StringUtils.isBlank(requestId)) {
                requestId = UUID.randomUUID().toString();
            }
            if (requestTrace == null) {
                requestTrace = Collections.emptyList();
            }
            return new RequestCtxData<>(
                requestId,
                Collections.unmodifiableList(requestTrace),
                locale,
                Collections.unmodifiableMap(contextAttributes),
                compAttsDisabled,
                msgLevel
            );
        }
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/request/RequestCtxData.kt
package ru.citeck.ecos.records3.record.request

import ecos.com.fasterxml.jackson210.databind.annotation.JsonDeserialize
import ru.citeck.ecos.records3.record.request.msg.MsgLevel
import java.util.*
import com.fasterxml.jackson.databind.annotation.JsonDeserialize as JackJsonDeserialize

@JsonDeserialize(builder = RequestCtxData.Builder::class)
@JackJsonDeserialize(builder = RequestCtxData.Builder::class)
data class RequestCtxData(
    val requestId: String,
    val requestTrace: List<String>,
    val locale: Locale,
    val ctxAtts: Map<String, Any?>,
    val computedAttsDisabled: Boolean,
    val msgLevel: MsgLevel,
    val omitErrors: Boolean,
    val readOnly: Boolean,
    val txnId: UUID?,
    val txnOwner: Boolean,
    val sourceIdMapping: Map<String, String>
) {

    companion object {

        @JvmStatic
        fun create(): Builder {
            return Builder()
        }

        @JvmStatic
        fun create(builder: Builder.() -> Unit): RequestCtxData {
            val builderObj = Builder()
            builder.invoke(builderObj)
            return builderObj.build()
        }
    }

    fun copy(): Builder {
        return Builder(this)
    }

    fun copy(builder: Builder.() -> Unit): RequestCtxData {
        val builderObj = Builder(this)
        builder.invoke(builderObj)
        return builderObj.build()
    }

    class Builder() {

        var requestId: String = ""
        var requestTrace = mutableListOf<String>()
        var locale: Locale = Locale.ENGLISH
        var ctxAtts: MutableMap<String, Any?> = mutableMapOf()
        var computedAttsDisabled = false
        var msgLevel: MsgLevel = MsgLevel.WARN
        var omitErrors: Boolean = false
        var readOnly: Boolean = false
        var txnId: UUID? = null
        var txnOwner: Boolean = false
        var sourceIdMapping: Map<String, String> = emptyMap()

        constructor(base: RequestCtxData) : this() {
            requestId = base.requestId
            requestTrace = base.requestTrace.toMutableList()
            locale = base.locale
            ctxAtts = base.ctxAtts.toMutableMap()
            computedAttsDisabled = base.computedAttsDisabled
            msgLevel = base.msgLevel
            omitErrors = base.omitErrors
            readOnly = base.readOnly
            txnId = base.txnId
            txnOwner = base.txnOwner
            sourceIdMapping = base.sourceIdMapping
        }

        fun withOmitErrors(omitErrors: Boolean): Builder {
            this.omitErrors = omitErrors
            return this
        }

        fun withRequestId(requestId: String): Builder {
            this.requestId = requestId
            return this
        }

        fun withRequestTrace(requestTrace: List<String>): Builder {
            this.requestTrace = requestTrace.toMutableList()
            return this
        }

        fun withLocale(locale: Locale): Builder {
            this.locale = locale
            return this
        }

        fun withCtxAtts(contextAtts: Map<String, Any?>): Builder {
            this.ctxAtts = contextAtts.toMutableMap()
            return this
        }

        fun withComputedAttsDisabled(disabled: Boolean): Builder {
            this.computedAttsDisabled = disabled
            return this
        }

        fun withMsgLevel(level: MsgLevel): Builder {
            this.msgLevel = level
            return this
        }

        fun withReadOnly(readOnly: Boolean?): Builder {
            this.readOnly = readOnly ?: false
            return this
        }

        fun withTxnId(txnId: UUID?): Builder {
            this.txnId = txnId
            return this
        }

        fun withTxnOwner(txnOwner: Boolean?): Builder {
            this.txnOwner = txnOwner ?: false
            return this
        }

        fun withSourceIdMapping(sourceIdMapping: Map<String, String>?): Builder {
            this.sourceIdMapping = sourceIdMapping ?: emptyMap()
            return this
        }

        fun build(): RequestCtxData {

            if (requestId.isBlank()) {
                requestId = UUID.randomUUID().toString()
            }

            return RequestCtxData(
                requestId,
                requestTrace,
                locale,
                ctxAtts,
                computedAttsDisabled,
                msgLevel,
                omitErrors,
                readOnly,
                txnId,
                txnOwner,
                sourceIdMapping
            )
        }
    }
}
